### PR TITLE
BigSet and ImageSetLabels/VideoSetLabels refactor (converted to Set subclasses)

### PR DESCRIPTION
**Pull request by benjaminpkane**
Updated.

`BigSet` with optional temp storage functionality.

`ImageSetLabels` and `VideoSetLabels` are now direct subclasses of `Set`.

`BigImageSetLabels` and `BigVideoSetLabels` implementations, as well.

### CODE REVIEW

File: eta/core/video.py [hunk 1]
```diff
@@ -1455,6 +1455,17 @@ def from_dict(cls, d):
         return cls(videos=videos, schema=schema)
 
 
+class _BigVideoLabelsSet(BigSet, _VideoLabelsSet):
+    '''Internal set type used by BigVideoSetLabels to store labels.'''
+
+
+class BigImageSetLabels(VideoSetLabels):
```
Comment (brimoor) [inline eta/core/video.py]: BigVideoSetLabels

File: eta/core/image.py [hunk 2]
```diff
@@ -786,6 +786,17 @@ def from_dict(cls, d):
         return cls(images=images, schema=schema)
 
 
+class _BigImageLabelsSet(BigSet, _ImageLabelsSet):
+    '''Internal set type used by BigImageSetLabels to store labels.'''
+
+
+class BigImageSetLabels(ImageSetLabels):
```
Comment (brimoor) [inline eta/core/image.py]: All the code finally pays off! Pretty slick that this is all we have to write to get a `Big` version of the class!

File: eta/core/image.py [hunk 3]
```diff
@@ -786,6 +786,17 @@ def from_dict(cls, d):
...
+    '''ImageSetLabels which uses a BigSet (_BigImageLabelsSet) to store labels.
```
Comment (brimoor) [inline eta/core/image.py]: No need to mention `_BigImageLabelsSet` explicitly. Users don't need to read about private classes.

I'd explicitly mention that `BigImageSetLabels` behaves identically to `ImageSetLabels` except that each element is stored on disk.

**Comment by brimoor** [comment]
@benjaminpkane feel free to introduce mixins if you see a good opportunity + value to do it. There are definitely opportunities. Even for `Container` and `Set`

File: eta/core/serial.py [hunk 4]
```diff
@@ -797,6 +797,382 @@ def _validate_dict(cls, d):
...
+            filters: a list of functions that accept elements and return
+                True/False
+            match: a function (usually `any` or `all`) that accepts an iterable
+                and returns True/False. Used to aggregate the outputs of each
+                filter to decide whether a match has occurred. The default is
+                `any`
+        '''
+        new_ele_set = self._filter_elements(filters, match)
+        keys = [key for key in self.__elements__ if key not in new_ele_set]
+        self.delete_inds(keys)
```
Comment (brimoor) [inline eta/core/serial.py]: `delete_keys`

**Comment by benjaminpkane** [comment]
After pondering some issues, I'd like to remove the `backing_dir` as a concept to the user and always use a self-managed temp directory for both `BigContainer` and `BigSet`. `write_json` would write out both the index and copy the element jsons to the index's base directory, and `from_json` would expect such a structure. Archiving would not change.

This work isn't critical to the sprint, so its not a priority. Come Monday this PR will be ready.

**Comment by brimoor** [comment]
Not necessarily opposed, but I'd be curious to hear your reasoning and what issues you encountered.

One halfway option is to give the user _optional_ control over the backing directory and default to the self-managed temp directory if they don't override it.

The copying of the backing directory to the index's base directory when serializing is the sticky part. I can imagine why you might be proposing that if you're trying to use `BigContainer` in an ETA pipeline. The pipeline system would need a way to automatically generate paths to work with BigContainers.

Perhaps `to_archive` should support an "uncompressed mode" where the data is just written to a directory without any archiving?

**Comment by benjaminpkane** [comment]
Most 

> Not necessarily opposed, but I'd be curious to hear your reasoning and what issues you encountered.
> 
> One halfway option is to give the user _optional_ control over the backing directory and default to the self-managed temp directory if they don't override it.
> 
> The copying of the backing directory to the index's base directory when serializing is the sticky part. I can imagine why you might be proposing that if you're trying to use `BigContainer` in an ETA pipeline. The pipeline system would need a way to automatically generate paths to work with BigContainers.
> 
> Perhaps `to_archive()` should support an "uncompressed mode" where the data is just written to a directory without any archiving?

Having the option might be better. My reasoning is that if `backing_dir` is required methods on `ImageLabels` (and the `VideoLabels` analogs) like `get_matches`, `count_matches`, `extract_image_labels`, and `copy` need to be overridden for `BigImageSetLabels` to accept a `backing_dir`.

It's not that much work, but seems like a troublesome practice for implementations in general. `BigContainer`s and `BigSet`s are also less interchangeable with the notion of a `backing_dir` also being present for any copy-like operations.

It might also be nice if `ImageSetLabels` was a direct subclass of `Set` (as opposed to the `images` attribute).

**Comment by brimoor** [comment]
I definitely agree that life would be much better if `ImageSetLabels` and `VideoSetLabels` were direct subclasses of `Set`! I didn't do it when I added `Set` because it will involve renaming some methods that are probably used elsewhere in our code.

But it would be waaaay better if this change were made, and `ImageSetLabels` and `VideoSetLabels` are relatively new, so I think now is the time to make this change :)

**Comment by benjaminpkane** [comment]
Sounds good! I will start with making `ImageSetLabels` and `VideoSetLabels` direct subclasses of `Set`.

**Comment by benjaminpkane** [comment]
The slicing PR has been merged into this branch.

`*SetLabels` are now direct subclasses of `Set`/`BigSet`.

The temp storage optional feature has only been added to `BigSet`. 

I've tested with `ImageSetLabels` and `BigImageSetLabels`, but no with their video counterparts. A few more things to do like `to_set` for `BigSet`.

**Pull request closed without merging**